### PR TITLE
MSVC Support

### PR DESCRIPTION
**Pull request by mnordie**
Hi Guys!
I've been trying to get Zipios to compile on windows.
I think I've come pretty far but I face some issues with the test suite.
On windows, for some reason, the zipfile.entries() doesn't contain entries for the directories causing the `REQUIRES(zf.size() == tree.size());` to fail.
Any insight on why this might be would be much appreciated!

### CODE REVIEW

**Comment by AlexisWilke** [comment]
That looks like a lot of changes to check... Have you tried to compile under Linux with all those changes?

Otherwise, I would imagine that if somehow the expected size is wrong, something does not work as expected. Where you able to check out the generated zip file(s) to see whether they include the expected number of entries? Under Linux we have a `zipinfo` tool to list the content of a zip file.

**Comment by dooglio** [comment]
Yes, please leave details of what you tested. And @AlexisWilke is right: this is a multi-platform tool, so testing under Linux is critical if we are to merge your PR.

File: tests/zipfile.cpp [hunk 1]
```diff
@@ -187,20 +188,73 @@ SCENARIO("ZipFile with a valid zip archive", "[ZipFile] [FileCollection]")
                 REQUIRE_FALSE(zf.getInputStream("inexistant", zipios::FileCollection::MatchPath::MATCH));
                 REQUIRE_FALSE(zf.getInputStream("inexistant", zipios::FileCollection::MatchPath::IGNORE));
                 REQUIRE(zf.getName() == "tree.zip");
-                REQUIRE(zf.size() == tree.size());
                 zf.mustBeValid(); // not throwing
-
+                if (zf.size() != tree.size())
```
Comment (mnordie) [inline tests/zipfile.cpp]: Debug code.
On windows these are not equal. tree.size() is larger than zf.size()
Seems it it the directory entries that are missing:
```
tree/f1
tree/f2
tree/d1/f4
tree/d1/f5
```
Whereas on linux you'd get

```
tree/f1
tree/f2
tree/d1   <--
tree/d1/f4
tree/d1/f5
```

**Comment by mnordie** [comment]
The test suite runs successfully on Linux.
Regarding the changes, they aren't that many actually, just many places ;)

- Added the `__declspec(dllexport)` to the public classes with the `ZIPIOSDLL_API` macro in order to link to Zipios dynamically.
- Use `vector.data()` instead of `&vector[0]` as the latter crashes on windows when vector is uninitialized.
- Initialize `zipios::g_separator` based on  os and use it instead of '/'

The only "real" change to the code is the addition of a `virtual void loadEntries() {}` to `FileCollection `in order to avoid creating a temporary copy of all the entries vectors when calling entries() only to make sure it's loaded. This made the unit tests complete 5 times faster. In debug mode, maybe this is is optimized away or otherwise not a problem for regular use. I've tested with and without this change.

So, the test. I frote a comment on the debug code in [https://github.com/Zipios/Zipios/pull/23/commits/01709f3adecd2b63650bc588fa6950a266936f76#diff-85e0610225e8b13b4bb63c9400d91b27](tests/zipfile.cpp)

**Comment by mnordie** [comment]
There is debug code here and there and stuff to clean up so I'm not proposing this is actually ready to be pulled, just wanted to get some feedback and help understanding the problems I'm facing.

**Comment by dooglio** [comment]
Hi @mnordie,

Thanks for your PR! 

Unfortunately, there are a *lot* of changes here. I think if you could split your changes up into separate PRs it might help to go faster.

For example, have one PR focused on adding the `__declspec(dllexport)` macros and nothing else. Then make sure CMake runs on both Windows and Linux.

Then focus on a PR which gets the library to compile under Windows. I would also like to see the focus on getting it to compile over style issues (like dropping the `struct` keyword off of a variable declaration, for example).

You can cherry pick your commits which deal exclusively with CMake issues and create a new branch for a separate PR. Just start there first, then we can roll in each piece as we go.

We don't have a full time team on this project, so if you can work with me on this, we are more likely to merge your work, which I want you to know, I really appreciate!

**Pull request closed without merging**